[SPARK-36107][SQL] Refactor first set of 20 query execution errors to use error classes

Peng-Lei opened the pull request:
### What changes were proposed in this pull request?
Refactor some exceptions in QueryExecutionErrors to use error classes. as follows:
```
columnChangeUnsupportedError
logicalHintOperatorNotRemovedDuringAnalysisError
cannotEvaluateExpressionError
cannotGenerateCodeForExpressionError
cannotTerminateGeneratorError
castingCauseOverflowError
cannotChangeDecimalPrecisionError
invalidInputSyntaxForNumericError
cannotCastFromNullTypeError
cannotCastError
cannotParseDecimalError
simpleStringWithNodeIdUnsupportedError
evaluateUnevaluableAggregateUnsupportedError
dataTypeUnsupportedError
dataTypeUnsupportedError
failedExecuteUserDefinedFunctionError
divideByZeroError
invalidArrayIndexError
mapKeyNotExistError
rowFromCSVParserNotExpectedError
```

### Why are the changes needed?
[SPARK-36107](https://issues.apache.org/jira/browse/SPARK-36107)

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existed UT Testcase

Peng-Lei committed:
use error classes to refact the first 20 errors in QueryExecutionErrors

AmplabJenkins commented:
Can one of the admins verify this patch?

Peng-Lei commented:
cc @karenfeng  FYI

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -63,5 +63,67 @@
+  },
+  "UNSUPPORTED_CHANGE_COLUMN" : {
+    "message" : [ "Please add an implementation for a column change here" ],
+    "sqlState" : "46110"
```
46110 is not in the SQLSTATE standard; I believe that `0A000` is the standard SQLSTATE for unsupported operations.

Peng-Lei replied:
done

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -63,5 +63,67 @@
+  },
+  "LOGICAL_HINT_OPERATOR_NOT_REMOVE_DURING_ANALYSIS" : {
+    "message" : [ "Internal error: logical hint operator should have been removed during analysis" ],
+    "sqlState" : "42000"
```
I don't think that internal errors count as syntax or semantic errors (42000); it may be simpler to remove the SQLSTATE entirely.

Peng-Lei replied:
done

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -63,5 +63,67 @@
+    "message" : [ "Internal error: logical hint operator should have been removed during analysis" ],
+    "sqlState" : "42000"
+  },
+  "CAN_NOT_EVALUATE_EXPRESSION" : {
```
Nit: I think `cannot` is more common than `can not`.

Peng-Lei replied:
done

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -63,5 +63,67 @@
+  },
+  "FAILED_EXECUTE_UDF" : {
+    "message" : [ "Failed to execute user defined function (%s: (%s) => %s)" ],
+    "sqlState" : "42000"
```
I'm not sure if 42000 is the right SQLSTATE here; maybe 45000 is a better fit.

Peng-Lei replied:
done

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -63,5 +63,67 @@
+    "message" : [ "Cannot parse any decimal" ],
+    "sqlState" : "42000"
+  },
+  "UNSUPPORTED_SIMPLESTRING_WITH_NODEID" : {
```
Nit: `SIMPLESTRING` -> `SIMPLE_STRING`

Peng-Lei replied:
done

karenfeng commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:
```diff
@@ -67,89 +68,116 @@ import org.apache.spark.util.CircularBuffer
 
   def columnChangeUnsupportedError(): Throwable = {
-    new UnsupportedOperationException("Please add an implementation for a column change here")
+    new UnsupportedOperationException(
```
Can you add a new exception type that mixes `UnsupportedOperationException` with `SparkThrowable`? It'd look like [`SparkArithmeticException`](https://github.com/apache/spark/blob/f90eb6a5db0778fd18b0b544f93eac3103bbf03b/core/src/main/scala/org/apache/spark/SparkException.scala#L75). Then we can catch these exceptions and use their error classes/SQLSTATEs.

Peng-Lei replied:
Of course, I'll do it. Thank you.

Peng-Lei replied:
@karenfeng I also add `SparkUnsupportedOperationException`, `SparkIllegalStateException`,`SparkNumberFormatException`,`SparkIllegalArgumentException`,`SparkArrayIndexOutOfBoundsException`,`SparkNoSuchElementException` that with `SparkThrowable`.

karenfeng commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:
```diff
@@ -67,89 +68,116 @@ import org.apache.spark.util.CircularBuffer
   }
 
   def logicalHintOperatorNotRemovedDuringAnalysisError(): Throwable = {
     new IllegalStateException(
```
Same thing here - we'll need to create a couple new exception classes.

karenfeng commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:
```diff
@@ -67,89 +68,116 @@ import org.apache.spark.util.CircularBuffer
 
   def castingCauseOverflowError(t: Any, targetType: String): ArithmeticException = {
-    new ArithmeticException(s"Casting $t to $targetType causes overflow")
+    new ArithmeticException(
```
You can use `SparkArithmeticException` here instead.

karenfeng reviewed (commented):
Thanks for working on this! We'll need to create some new `SparkThrowable` implementations; if we simply throw the existing exception types, the error class and SQLSTATE are not propagated through.

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -63,5 +63,67 @@
     "message" : [ "Writing job aborted" ],
     "sqlState" : "40000"
+  },
+  "UNSUPPORTED_CHANGE_COLUMN" : {
```
These need to be in alphabetical order to pass the style tests.

Peng-Lei replied:
done

Peng-Lei committed:
Modify the code based on the review comments

1. add SparkUnsupportedOperationException
       SparkIllegalStateException
       SparkNumberFormatException
       SparkIllegalArgumentException
       SparkArrayIndexOutOfBoundsException
       SparkNoSuchElementException
2. change the sqlstate suitable code and order the error in
   error-class.json
3. others

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -3,6 +3,34 @@
+  },
+  "CANNOT_CHANGE_DECIMAL_PRECISION" : {
+    "message" : [ "%s cannot be represented as Decimal(%s, %s)." ],
+    "sqlState" : "HY104"
```
I believe that HY represents CLI-specific conditions. This seems to be triggered by overflows; maybe 22023 is a better fit.

Peng-Lei replied:
Yeah. Maybe 22023 is a better fit. The exception is triggered by invalid parameter value.

karenfeng reviewed (approved):
Thanks for doing this! LGTM. @HyukjinKwon, can you also take a look?

Peng-Lei commented:
@HyukjinKwon Could you take a look?

Peng-Lei committed:
Merge branch 'master' into SPARK-36017

Peng-Lei commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -14,6 +42,10 @@
   },
+  "FAILED_EXECUTE_UDF" : {
+    "message" : [ "Failed to execute user defined function (%s: (%s) => %s)" ],
+    "sqlState" : "45000"
```
The SQLSTATE '45000' is not valid. Because it is not included in [README](https://github.com/apache/spark/blob/master/core/src/main/resources/error/README.md)

Peng-Lei replied:
@karenfeng I delete it.

Peng-Lei committed:
Recheck the SQLSTATE to ensure it is valid.

Peng-Lei committed:
mark the exception as private[spark]

Peng-Lei committed:
Merge branch 'master' into SPARK-36017

HyukjinKwon commented:
Sorry I saw this late. There have been few changes made. @karenfeng mind double checking before we go ahead please?

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -3,6 +3,34 @@
+    "sqlState" : "42000"
+  },
+  "CANNOT_PARSE_DECIMAL" : {
+    "message" : [ "Cannot parse any decimal" ],
```
This is a strange message. Can we change this to `Cannot parse decimal`?

Peng-Lei replied:
ok

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -3,6 +3,34 @@
+    "sqlState" : "22023"
+  },
+  "CANNOT_EVALUATE_EXPRESSION" : {
+    "message" : [ "Cannot evaluate expression: %s: %s" ],
```
This is a little awkward in the case that there is no value for the first param (it would look like `Cannot evaluate expression: : expression`).

I think we should just have it be: `Cannot evaluate expression: %s` and pass `methodName: unEvaluable.toString` as the single param.

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -3,6 +3,34 @@
+  },
+  "CANNOT_EVALUATE_EXPRESSION" : {
+    "message" : [ "Cannot evaluate expression: %s: %s" ],
+    "sqlState" : "42000"
```
This is likely an internal error; I think we can remove the SQLSTATE here.

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -96,10 +144,21 @@
     "message" : [ "Unsupported literal type %s %s" ],
     "sqlState" : "0A000"
   },
+  "UNSUPPORTED_SIMPLE_STRING_WITH_NODEID" : {
```
Nit: `UNSUPPORTED_SIMPLE_STRING_WITH_NODEID` -> `UNSUPPORTED_SIMPLE_STRING_WITH_NODE_ID`

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -54,9 +88,19 @@
+  "LOGICAL_HINT_OPERATOR_NOT_REMOVE_DURING_ANALYSIS" : {
+    "message" : [ "Internal error: logical hint operator should have been removed during analysis" ]
+  },
+  "MAP_KEY_NOT_EXIST" : {
```
The grammar here is a bit clumsy. Can we make this `MAP_KEY_DOES_NOT_EXIST`?

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -96,10 +144,21 @@
+    "sqlState" : "0A000"
+  },
+  "UNSUPPORTED_DATATYPE" : {
+    "message" : [ "Unsupported data type %s %s" ],
```
This is a little awkward in the case that there is no value for the first param; as discussed above, I think we should combine the two-param case into one string param.

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -54,9 +88,19 @@
     "message" : [ "The fraction of sec must be zero. Valid range is [0, 60]." ],
     "sqlState" : "22023"
   },
+  "INVALID_INPUT_SYNTAX_FOR_NUMERIC" : {
```
Nit: `INVALID_INPUT_SYNTAX_FOR_NUMERIC` -> `INVALID_INPUT_SYNTAX_FOR_NUMERIC_TYPE`

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -3,6 +3,34 @@
+    "message" : [ "Cannot terminate expression: %s" ],
+    "sqlState" : "42000"
+  },
+  "CAST_CAUSE_OVERFLOW" : {
```
Nit: `CAST_CAUSE_OVERFLOW` -> `CAST_CAUSES_OVERFLOW`

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -3,6 +3,34 @@
+  },
+  "CAST_CAUSE_OVERFLOW" : {
+    "message" : [ "Casting %s to %s causes overflow" ],
+    "sqlState" : "22018"
```
I think `22005` is a better fit here.

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -3,6 +3,34 @@
+  },
+  "CANNOT_GENERATE_CODE_FOR_EXPRESSION" : {
+    "message" : [ "Cannot generate code for expression: %s" ],
+    "sqlState" : "42000"
```
I'm not sure this has an obvious SQLSTATE.

karenfeng commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -3,6 +3,34 @@
+  },
+  "CANNOT_TERMINATE_GENERATOR" : {
+    "message" : [ "Cannot terminate expression: %s" ],
+    "sqlState" : "42000"
```
I'm not sure this has an obvious SQLSTATE.

karenfeng reviewed (commented):
Thanks for working on this @Peng-Lei! I left some comments - please tag me once you address them.

Peng-Lei committed:
Merge branch 'apache:master' into SPARK-36017

Peng-Lei committed:
address comments

Peng-Lei commented:
@karenfeng I've already address the comments. Please check. Thank you.

karenfeng reviewed (approved):
LGTM, thanks @Peng-Lei! @HyukjinKwon, can you help take a second look?

HyukjinKwon commented:
Merged to master.

Peng-Lei commented:
@karenfeng @HyukjinKwon Thank you！

MaxGekk commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -3,6 +3,31 @@
     "message" : [ "Field name %s is ambiguous and has %s matching fields in the struct." ],
     "sqlState" : "42000"
   },
+  "CANNOT_CAST_DATATYPE" : {
```
Any ideas how to trigger this error class? I haven't found any ways how trigger it from user code. WDYT? @cloud-fan @HyukjinKwon
